### PR TITLE
Error on tangent

### PR DESCRIPTION
**Pull request by GaelVaroquaux**
ENH: error for a single subject in fit_transform for tangent space

A classic error is that people call fit_transform on single subjects when the tangent space parametrization. This cannot work.

This PR raises an error in this case.

The code is extracted from PR #1477 that we will probably drop.

### CODE REVIEW

**Comment by KamalakerDadi** [comment]
>The code is extracted from PR #1477 that we will probably drop.

Why will it be dropped ?

**Comment by GaelVaroquaux** [comment]
Because it doesn't seem to make an improvement. 

⁣Sent from my phone. Please forgive typos and briefness.​

File: nilearn/connectome/connectivity_matrices.py [hunk 1]
```diff
@@ -466,20 +451,57 @@ def fit(self, X, y=None):
...
+        self.cov_estimator_ = clone(self.cov_estimator)
 
+        covariances = [self.cov_estimator_.fit(x).covariance_ for x in X]
         if self.kind == 'tangent':
-            covariances = [self.cov_estimator_.fit(x).covariance_ for x in X]
             self.mean_ = _geometric_mean(covariances, max_iter=30, tol=1e-7)
             self.whitening_ = _map_eigenvalues(lambda x: 1. / np.sqrt(x),
                                                self.mean_)
+        else:
+            self.mean_ = np.mean(covariances, axis=0)
```
Comment (salma1601) [inline nilearn/connectome/connectivity_matrices.py]: Hi ! 
Are you retrurning the covariances mean even for kind='correlation' or précision ?

**Comment by GaelVaroquaux** [comment]
> Are you retrurning the covariances mean even for kind='correlation' or
> précision ?

Hum, good point. Do you think that we should return the mean "kind"?

File: nilearn/connectome/connectivity_matrices.py [hunk 2]
```diff
@@ -466,20 +451,57 @@ def fit(self, X, y=None):
...
+        if self.kind in ('tangent', 'tangent_geometric'):
+            # Check that people are applying fit_transform to a group of
+            # subject
+            # We can only impose this in fit_transform, as it is legit to
+            # fit only on a single given reference point
+            if not len(X) > 1:
+                raise ValueError("Tangent space parametrization can only "
+                    "be applied to a group of subjects, as it returns "
+                    "deviations to the mean. You provided %r" % X
+                    )
```
Comment (bthirion) [inline nilearn/connectome/connectivity_matrices.py]: I'm missing something. IIUC fit() accepts a single subject, and so does transform(). Only fit_transform() doesn't. But actually fit_transform() on a single subject could yield a null matrix.
?

**Comment by salma1601** [comment]
>Do you think that we should return the mean "kind"?

Yes, I think the user would expect the mean_ to correspond to the "kind" he has chosen

File: nilearn/connectome/connectivity_matrices.py (hunk 2 above)
Comment (salma1601) [reply nilearn/connectome/connectivity_matrices.py]: BTW, why is it an error to get a null matrix ? All the deviations are somehow "strange" matrices for a user that didn't understand that they are deviations from the mean: they can be négative on the diagonal, have values outside [-1, 1] etc. May be à simple warning is sufficient, to remind the context. May be even à message printed systematically for the tangent kind would be pedagogic, whatever the number of subjects is.

**Comment by GaelVaroquaux** [comment]
> I'm missing something. IIUC fit() accepts a single subject, and so does
> transform(). Only fit_transform() doesn't. But actually fit_transform() on a
> single subject could yield a null matrix.
> ?

The problem is that fit defines the reference point, and transform
computes the delta to the reference point. So it is not possible to have
a single subject for fit_transform. It's a common mistake: people do
that, and don't understand what they are doing wrong.

The docs could probably be improved :)

**Comment by GaelVaroquaux** [comment]
> BTW, why is it an error to get a null matrix ?

It's something that's not useful.

The problem is that users who don't understand what they are doing call
'fit_transform' in a loop on each subject of a group and expect to do a
group analysis on the result. That does not work.

**Comment by salma1601** [comment]
>The problem is that users who don't understand what they are doing call 'fit_transform' in a loop on each subject of a group and expect to do a group analysis on the result. That does not work.

Ah ok now I see, thank you. You are right.
Something that always worried me about `tangent` is that people comparing groups may compute their connectivities for each group separately and then do statistics/classification. This also will be wrong. I already saw someone doing it and don't know how to prevent this. Doc ?

File: nilearn/connectome/connectivity_matrices.py [hunk 3]
```diff
@@ -466,20 +451,57 @@ def fit(self, X, y=None):
...
+        else:
+            self.mean_ = np.mean(covariances, axis=0)
+            # Fight numerical instabilities: make symmetric
+            self.mean_ = self.mean_ + self.mean_.T
+            self.mean_ *= .5
 
         return self
 
+    def fit_transform(self, X, y=None):
+        if self.kind in ('tangent', 'tangent_geometric'):
```
Comment (salma1601) [inline nilearn/connectome/connectivity_matrices.py]: do we actually have `tangent_geometric` and `tangent` ?

**Comment by GaelVaroquaux** [comment]
> I already saw someone doing it and don't know how to prevent this. Doc ?

I think that doc and examples are probably the best solution. Though this
PR is an attempt to raise awareness.

**Comment by GaelVaroquaux** [comment]
> do we actually have tangent_geometric and tangent ?

Good catch! This was a mistake. Fixed!

**Comment by GaelVaroquaux** [comment]
I need a new review on this: I have made changes according to the suggestion of @salma1601 so that mean_ is not the mean of the corresponding object (ie mean of partial correlation, is kind='partial_correlation').

**Comment by salma1601** [comment]
Hi !
What happens if we first do fit and then transform separately with kind='tangent' ?

**Comment by salma1601** [comment]
Looks like the errir is not raised in this case ?

**Comment by GaelVaroquaux** [comment]
> What happens if we first do fit and then transform separately with kind=
> 'tangent' ?

You're somewhat breaking the contract, there, aren't you?

**Comment by bthirion** [comment]
I'm happy with the latest change. Shouldn't it be showcased in an example ?

**Comment by salma1601** [comment]
>You're somewhat breaking the contract, there, aren't you?

Oh, sorry, I wasn't aware this shouldn't be done ! I though we can either do fit_transform or first fit then transform...
OK, in this case LGTM.

**Comment by GaelVaroquaux** [comment]
Merging this guy.

**Pull request merged**